server/src/main/java/org/elasticsearch/common/lucene/Lucene.java

```diff
@@ -816,6 +798,63 @@ public int length() {
         };
     }
 
+    /**
+     * Return a {@link DocIdSetIterator} view of the provided scorer.
+     * @see #asDocIdSetIterator(int, ScorerSupplier, long)
+     */
+    public static DocIdSetIterator asDocIdSetIterator(final int maxDoc, @Nullable ScorerSupplier scorerSupplier) throws IOException {
+        return asDocIdSetIterator(maxDoc, scorerSupplier, 0L);
+    }
+
+    /**
+     * Given a {@link ScorerSupplier}, return a {@link DocIdSetIterator} instance that iterates over
+     * all documents contained in the set.
+     * @param estimatedGetCount an estimation of the number of times that {@link Scorer} will get called
+     */
+    public static DocIdSetIterator asDocIdSetIterator(final int maxDoc, @Nullable ScorerSupplier scorerSupplier, long estimatedGetCount)
+        throws IOException {
+        if (scorerSupplier == null) {
+            return DocIdSetIterator.empty();
+        }
+        final Scorer scorer = scorerSupplier.get(estimatedGetCount); // this never returns null
+        final TwoPhaseIterator twoPhase = scorer.twoPhaseIterator();
+        final DocIdSetIterator iterator = (twoPhase != null) ? twoPhase.approximation() : scorer.iterator();
+
+        return new DocIdSetIterator() {
```

Comment:
FYI Lucene already has this via `TwoPhaseIterator#asDocIdSetIterator`. That said, I don't think we should use it because it defeats the point of using a `TwoPhaseIterator`? For instance, if the main query matches doc IDs 3 and 10 only, and we call `advance(3)` on the filter of the `filter` aggregation: if 3 doesn't match, it will then try out 4 then 5... until it finds one that matches, but we don't care about them since they don't match the query?

Reply:
Ideally, the competitive aggregator would ignore the `TwoPhaseIterator` and just look at the approximation part, so that `TwoPhaseIterator#matches` would only be called when both the main query and the filter match.

Reply:
Not exactly. I pushed a small change that shows what I have in mind athttps://github.com/elastic/elasticsearch/compare/main...jpountz:elasticsearch:filter_competitive_iterator_poc?expand=1. Beware it's not tested and has at least one bug, but hopefully it helps?

Reply:
Nice, this is much cleaner. Ptal